grammers-session/src/peer.rs

```diff
@@ -255,138 +255,227 @@ impl fmt::Display for PeerId {
 }
 
 impl From<PeerInfo> for PeerRef {
+    #[inline]
     fn from(peer: PeerInfo) -> Self {
-        PeerRef {
+        <Self as From<&PeerInfo>>::from(&peer)
+    }
+}
+impl<'a> From<&'a PeerInfo> for PeerRef {
+    fn from(peer: &'a PeerInfo) -> Self {
+        Self {
             id: peer.id(),
             auth: peer.auth(),
         }
     }
 }
 
 impl From<tl::enums::Peer> for PeerId {
+    #[inline]
     fn from(peer: tl::enums::Peer) -> Self {
+        <Self as From<&tl::enums::Peer>>::from(&peer)
+    }
+}
+impl<'a> From<&'a tl::enums::Peer> for PeerId {
+    fn from(peer: &'a tl::enums::Peer) -> Self {
+        use tl::enums::Peer;
         match peer {
-            tl::enums::Peer::User(user) => PeerId::from(user),
-            tl::enums::Peer::Chat(chat) => PeerId::from(chat),
-            tl::enums::Peer::Channel(channel) => PeerId::from(channel),
+            Peer::User(user) => Self::from(user),
+            Peer::Chat(chat) => Self::from(chat),
+            Peer::Channel(channel) => Self::from(channel),
         }
     }
 }
 
 impl From<tl::types::PeerUser> for PeerId {
+    #[inline]
     fn from(user: tl::types::PeerUser) -> Self {
-        PeerId::user(user.user_id)
+        <Self as From<&tl::types::PeerUser>>::from(&user)
+    }
+}
+impl<'a> From<&'a tl::types::PeerUser> for PeerId {
+    fn from(user: &'a tl::types::PeerUser) -> Self {
+        Self::user(user.user_id)
     }
 }
 
 impl From<tl::types::PeerChat> for PeerId {
-    fn from(chat: tl::types::PeerChat) -> Self {
-        PeerId::chat(chat.chat_id)
+    #[inline]
+    fn from(user: tl::types::PeerChat) -> Self {
+        <Self as From<&tl::types::PeerChat>>::from(&user)
+    }
+}
+impl<'a> From<&'a tl::types::PeerChat> for PeerId {
+    fn from(chat: &'a tl::types::PeerChat) -> Self {
+        Self::chat(chat.chat_id)
     }
 }
 
 impl From<tl::types::PeerChannel> for PeerId {
+    #[inline]
     fn from(channel: tl::types::PeerChannel) -> Self {
-        PeerId::channel(channel.channel_id)
+        <Self as From<&tl::types::PeerChannel>>::from(&channel)
+    }
+}
+impl<'a> From<&'a tl::types::PeerChannel> for PeerId {
```

Comment:
These are nice to have (and I can see why people would use macros), but it's again unfortunate this only works for one reference type.

Reply:
What other types are you referring to by “only one reference type”?

Reply:
That's what [`Deref`](<https://doc.rust-lang.org/core/ops/trait.Deref.html>) is for.